### ISSUE TITLE
How to properly read a POST request?

### ISSUE THREAD

**Issue by n-gram-hub** [benoitc/gunicorn#2697]
Hi, I'm trying Gunicorn for the first time. 

I'm having lots of troubles when reading POST data: I must simply read some values coming from a PHP with a basic cURL request.

```php
$post = [
    'username' => 'user1',
    'password' => 'passuser1',
    'gender'   => 1,
];

$ch = curl_init('http://...');
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch, CURLOPT_POSTFIELDS, $post);

$response = curl_exec($ch);

curl_close($ch);
```

My code is very simple, but I don't understand how to get my three post values from _environ_ in Python.

Could you help me? I could not find any working example.

If this is not the place to ask this kind of question, please let me know. 

Thanks in advance.

**Comment by jamadden** [comment]
I don't think this is the right place to ask the question. You seem to be asking about the basics of Python's [WSGI](https://www.python.org/dev/peps/pep-3333/) specification, which is independent of gunicorn or any other WSGI server that implements the specification. In general, WSGI servers like gunicorn aren't concerned about the contents of an HTTP request body, and it's up to you to do something useful with it in your WSGI callable (for example, you may use functions from the standard library [`cgi` module](https://docs.python.org/3/library/cgi.html) to perform parsing).

Most people don't directly write WSGI callables or deal directly with parsing HTTP requests, however. They use some sort of library or framework to do that for them. One of the simplest that deals just with HTTP requests and responses directly is [WebOb](https://docs.pylonsproject.org/projects/webob/en/stable/), specifically `webob.request.Request`:
```python
def wsgi_callable(environ, start_response):
   request = webob.request.Request(environ)
   if 'some_key' in request.POST:
      …
```
There are numerous other frameworks that provide things like request routing based on the incoming path, and so on, thinks like Pyramid and Flask and many others.

**Comment by n-gram-hub** [comment]
@jamadden Hi, and thanks for your kind -and helpful- reply. Problem is I generally don't do any "server Python" and, at the same time, I must build something really simple for r&d, not production. I didn't suspect I needed to use a framework to simply read a post request. For example, in a very limited (if compared to Python) language like PHP, it's a matter of reading the content of $['POST']. Of course, I know PHP was created specifically with this kind of thing in mind. Nonetheless, I hoped Python/WSGI would be easier to work with. 

Anyway, to read the only variable I was receiving, I had to do the following:

```
 request_body = environ['wsgi.input'].read().decode()
 myData = urllib.parse.parse_qs(request_body, True)
 myField = myData[' name'][0].split("\n")[2]
```

It's horrible, but it works. I'll check into any decent framework if this little project should become anything serious.   :-)

Sorry for asking in the wrong place!
